pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+		// First get the group with all labels
+		b = series.Labels.Bytes(b)
+		g, groupExists := groups[string(b)]
+		if !groupExists {
+			lb := labels.NewBuilder(series.Labels)
+			g.labels = lb.Labels()
+			g.group = bucketGroupPool.Get()
+			g.group.groupedMetricName = series.Labels.Get(labels.MetricName)
+			g.group.pointBuckets = make([]buckets, h.timeRange.StepCount)
+			groups[string(b)] = g
+		}
+		g.group.remainingSeriesCount++
+		seriesGroupPair[0] = g.group
+
+		// Now, ignore `le` and create a group with the rest of the labels.
+		// We do this even if we don't have an `le` label. This is so we can detect
+		// where native histograms exist alongside classic histograms.
+		// So seriesGroupPair[0] could be the same as seriesGroupPair[1]
+		b = series.Labels.BytesWithoutLabels(b, labels.BucketLabel)
+		g, groupExists = groups[string(b)]
+
+		if !groupExists {
+			lb := labels.NewBuilder(series.Labels)
+			lb.Del(labels.BucketLabel)
+			g.labels = lb.Labels()
+			g.group = bucketGroupPool.Get()
+			g.group.groupedMetricName = series.Labels.Get(labels.MetricName)
+			g.group.pointBuckets = make([]buckets, h.timeRange.StepCount)
+			groups[string(b)] = g
+		}
+		g.group.remainingSeriesCount++
+		seriesGroupPair[1] = g.group
+		h.seriesGroups = append(h.seriesGroups, seriesGroupPair)
+	}
+
+	seriesMetadata := types.GetSeriesMetadataSlice(len(groups))
+	h.remainingGroups = make([]*bucketGroup, 0, len(groups))
+	for _, g := range groups {
+		seriesMetadata = append(seriesMetadata, types.SeriesMetadata{Labels: g.labels.DropMetricName()})
+		h.remainingGroups = append(h.remainingGroups, g.group)
+	}
+
+	return seriesMetadata, nil
+}
+
+func (h *HistogramFunctionOverInstantVector) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	if len(h.remainingGroups) == 0 {
+		// No more groups left.
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	// Determine next group to return
+	thisGroup := h.remainingGroups[0]
+	h.remainingGroups = h.remainingGroups[1:]
+	defer func() {
+		// Reset the group before returning to the pool
+		thisGroup.groupedMetricName = ""
+		thisGroup.pointBuckets = nil
+		thisGroup.nativeHistograms = nil
+		thisGroup.remainingSeriesCount = 0
+		bucketGroupPool.Put(thisGroup)
+	}()
+
+	// Iterate through inner series until the desired group is complete
+	if err := h.accumulateUntilGroupComplete(ctx, thisGroup); err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	if len(h.remainingGroups) == 0 {
+		// We no longer need the bucketValues strings
+		types.StringSlicePool.Put(h.bucketValues, h.memoryConsumptionTracker)
+	}
+
+	return h.computeOutputSeriesForGroup(ctx, thisGroup)
+}
+
+// accumulateUntilGroupComplete gathers all the series associated with the given bucketGroup
+// As each inner series is selected, it is added into its respective groups.
+// This means a group other than the one we are focused on may get completed first, but we
+// continue until the desired group is ready.
```

Comment:
We should order the groups so this never happens - see comment above about sorting the output series.